invoice/src/main/java/org/killbill/billing/invoice/InvoicePluginDispatcher.java

```diff
@@ -173,6 +182,52 @@ private void onCompletionCall(final boolean isSuccess,
         }
     }
 
+    public List<DefaultInvoice> splitInvoices(final DefaultInvoice originalInvoice,
+                                              final boolean isDryRun,
+                                              final CallContext callContext,
+                                              // The pluginProperties list passed to plugins is mutable by the plugins
+                                              @SuppressWarnings("TypeMayBeWeakened") final LinkedList<PluginProperty> properties,
+                                              final InternalTenantContext tenantContext) {
+        log.debug("Invoking invoice plugins for splitInvoices operation: isDryRun='{}', originalInvoice='{}'", isDryRun, originalInvoice);
+
+        final Collection<InvoicePluginApi> invoicePlugins = getInvoicePlugins(tenantContext).values();
+        final Invoice clonedInvoice = (Invoice) originalInvoice.clone();
+        for (final InvoicePluginApi invoicePlugin : invoicePlugins) {
+            final InvoiceGroupingResult grpResult = invoicePlugin.getInvoiceGrouping(clonedInvoice, isDryRun, properties, callContext);
+            if (grpResult != null && grpResult.getInvoiceGroups() != null && grpResult.getInvoiceGroups().size() > 0) {
+
+                final List<DefaultInvoice> result = new ArrayList<>();
+                final Map<UUID, InvoiceItem> itemMap = originalInvoice.getInvoiceItems()
+                                                                      .stream()
+                                                                      .map(new Function<InvoiceItem, AbstractMap.SimpleEntry<UUID, InvoiceItem>>() {
+                                                                          @Override
+                                                                          public SimpleEntry<UUID, InvoiceItem> apply(final InvoiceItem invoiceItem) {
+                                                                              return new SimpleEntry<>(invoiceItem.getId(), invoiceItem);
+                                                                          }
+                                                                      }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+                final List<InvoiceGroup> groups = grpResult.getInvoiceGroups();
+                for (final InvoiceGroup grp : groups) {
+                    final DefaultInvoice grpInvoice = (DefaultInvoice) originalInvoice.clone();
+                    final UUID newInvoiceId = UUIDs.randomUUID();
+                    grpInvoice.setId(newInvoiceId);
```

Comment:
Nit: maybe we could introduce a `cloneForGrouping` which hides that cleanup.

Reply:
I went back and forth and decided to simply reuse a [CTOR that already exists](https://github.com/killbill/killbill/pull/1687/commits/c133922b00dbe10ee89cd5b6e6897299d0719b7c#diff-a6f829e33b0216520626dd03b04a2a7ac9c9018a7e39bbafb09e4e9c36de0c60R209).